File: frontend/types/notification.ts
/**
 * Notification type definitions matching backend API schemas
 */

export type NotificationChannel = 'email' | 'in_app' | 'webhook';
export type NotificationPriority = 'low' | 'normal' | 'high' | 'urgent';
export type NotificationEventType =
  | 'project_created'
  | 'project_updated'
  | 'project_deleted'
  | 'baseline_created'
  | 'baseline_deviation'
  | 'sprint_started'
  | 'sprint_completed'
  | 'task_assigned'
  | 'task_completed'
  | 'deadline_approaching'
  | 'deadline_missed';

export interface Notification {
  id: string;
  user_id: string;
  title: string;
  message: string;
  event_type: NotificationEventType;
  priority: NotificationPriority;
  channel: NotificationChannel;
  is_read: boolean;
  metadata?: Record<string, any>;
  created_at: string;
  read_at?: string;
}

export interface NotificationRule {
  id: string;
  user_id: string;
  name: string;
  event_type: NotificationEventType;
  channels: NotificationChannel[];
  enabled: boolean;
  conditions?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface NotificationRuleCreate {
  name: string;
  event_type: NotificationEventType;
  channels: NotificationChannel[];
  enabled?: boolean;
  conditions?: Record<string, any>;
}

export interface NotificationRuleUpdate {
  name?: string;
  event_type?: NotificationEventType;
  channels?: NotificationChannel[];
  enabled?: boolean;
  conditions?: Record<string, any>;
}

export interface NotificationLog {
  id: string;
  notification_id: string;
  channel: NotificationChannel;
  status: 'pending' | 'sent' | 'failed' | 'delivered';
  error_message?: string;
  sent_at?: string;
  delivered_at?: string;
  created_at: string;
}

export interface NotificationTemplate {
  id: string;
  name: string;
  event_type: NotificationEventType;
  subject_template: string;
  body_template: string;
  channel: NotificationChannel;
  created_at: string;
  updated_at: string;
}

export interface NotificationListResponse {
  total: number;
  limit: number;
  offset: number;
  notifications: Notification[];
}

export interface NotificationRuleListResponse {
  total: number;
  rules: NotificationRule[];
}

export interface NotificationLogListResponse {
  total: number;
  limit: number;
  offset: number;
  logs: NotificationLog[];
}

export interface NotificationStats {
  unread_count: number;
  total_count: number;
  by_priority: Record<NotificationPriority, number>;
  by_event_type: Record<NotificationEventType, number>;
}

File: backend/app/schemas/notification.py
"""Pydantic schemas for notification API."""

from datetime import datetime
from typing import Optional, List, Dict, Any
from uuid import UUID

from pydantic import BaseModel, Field


# Notification schemas
class NotificationBase(BaseModel):
    """Base notification schema."""
    type: str
    title: str
    message: str
    metadata: Optional[Dict[str, Any]] = None


class NotificationCreate(NotificationBase):
    """Schema for creating a notification."""
    user_id: UUID


class NotificationUpdate(BaseModel):
    """Schema for updating a notification."""
    status: Optional[str] = None


class NotificationResponse(NotificationBase):
    """Schema for notification responses."""
    id: UUID
    user_id: UUID
    status: str
    read_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class NotificationListResponse(BaseModel):
    """Schema for notification list responses."""
    notifications: List[NotificationResponse]
    total: int
    limit: int
    offset: int


# Notification Rule schemas
class NotificationRuleBase(BaseModel):
    """Base notification rule schema."""
    event_type: str
    enabled: bool = True
    channels: List[str] = Field(default_factory=lambda: ["in_app"])
    conditions: Optional[Dict[str, Any]] = None


class NotificationRuleCreate(NotificationRuleBase):
    """Schema for creating a notification rule."""
    pass


class NotificationRuleUpdate(BaseModel):
    """Schema for updating a notification rule."""
    enabled: Optional[bool] = None
    channels: Optional[List[str]] = None
    conditions: Optional[Dict[str, Any]] = None


class NotificationRuleResponse(NotificationRuleBase):
    """Schema for notification rule responses."""
    id: UUID
    user_id: UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class NotificationRuleListResponse(BaseModel):
    """Schema for notification rule list responses."""
    rules: List[NotificationRuleResponse]
    total: int


# Event trigger schema
class NotificationEventTrigger(BaseModel):
    """Schema for triggering notification events."""
    event_type: str
    event_data: Dict[str, Any] = Field(
        description="Event data including title, message, and metadata"
    )


class NotificationEventResponse(BaseModel):
    """Schema for event trigger responses."""
    status: str
    event_type: str
    rules_matched: int
    notifications_created: int
    emails_queued: int
